src/cargo/util/auth/mod.rs

```diff
@@ -1,142 +1,106 @@
-    where
-        F: FnOnce(T) -> U,
-    {
-        Secret::from(f(self.inner))
-    }
+use crate::util::config::Value;
+use crate::util::credential::adaptor::BasicProcessCredential;
+use crate::util::credential::paseto::PasetoCredential;
+
+use super::{
+    config::{CredentialCacheValue, PathAndArgs},
+    credential::process::CredentialProcessCredential,
+    credential::token::TokenCredential,
+};
+
+#[derive(Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct RegistryConfig {
+    pub index: Option<String>,
+    pub token: Option<Secret<String>>,
+    pub credential_provider: Option<PathAndArgs>,
+    pub secret_key: Option<Secret<String>>,
+    pub secret_key_subject: Option<String>,
+    #[serde(rename = "default")]
+    _default: Option<String>,
+    #[serde(rename = "protocol")]
+    _protocol: Option<String>,
+    #[serde(rename = "credential-providers")]
+    _credential_providers: Option<Vec<String>>,
 }
 
-impl<T: ToOwned + ?Sized> Secret<&T> {
-    /// Converts a `Secret` containing a borrowed type to a `Secret` containing the
-    /// corresponding owned type.
-    /// ```
-    /// # use cargo::util::auth::Secret;
-    /// let borrowed: Secret<&str> = Secret::from("token");
-    /// let owned: Secret<String> = borrowed.owned();
-    /// ```
-    pub fn owned(&self) -> Secret<<T as ToOwned>::Owned> {
-        Secret::from(self.inner.to_owned())
-    }
-}
+/// Get the list of credential providers for a registry source.
+fn credential_provider(config: &Config, sid: &SourceId) -> CargoResult<Vec<Vec<String>>> {
+    let cfg = registry_credential_config_raw(config, sid)?;
+    let allow_cred_proc = config.cli_unstable().credential_process;
+    let providers = match cfg {
+        // Error for both `token` and `secret-key`
+        Some(RegistryConfig {
+            token: Some(_),
+            secret_key: Some(_),
+            ..
+        }) => {
+            return Err(format_err!(
+                "both `token` and `secret-key` \
+                were specified in the config for {sid}.\n\
+                Only one of these values may be set, remove one or the other to proceed.",
+            ))
+        }
 
-impl<T, E> Secret<Result<T, E>> {
-    /// Converts a `Secret<Result<T, E>>` to a `Result<Secret<T>, E>`.
-    pub fn transpose(self) -> Result<Secret<T>, E> {
-        self.inner.map(|v| Secret::from(v))
-    }
-}
+        // If there's a specific provider configured, use it.
+        Some(RegistryConfig {
+            credential_provider: Some(provider),
+            ..
+        }) if allow_cred_proc => vec![resolve_credential_alias(config, provider)],
 
-impl<T: AsRef<str>> Secret<T> {
-    /// Checks if the contained value is empty.
-    pub fn is_empty(&self) -> bool {
-        self.inner.as_ref().is_empty()
-    }
-}
+        // If a token is configured, use the token provider.
+        Some(RegistryConfig { token: Some(_), .. }) => vec![vec!["cargo:token".to_string()]],
```

Comment:
One of my concerns about how this fallback works is the following scenario:

1. Someone has `registry.token` or `registries.NAME.token` saved in the credentials.toml.
2. They want to switch to using a credential provider, so they set `registry.credential-providers = ["…"]`
3. They run some commands, but the provider never works, it only uses the old token.

What do you think about doing the following? In this case, check if the fallback providers contains `cargo:token`. If it does, proceed using the fallback list of providers. If it does not...I'm not sure. I didn't like the old error message. Perhaps it could just ignore the token? Or generate a warning that the token is ignored, and indicate the file where you can delete it to remove the warning?

Reply:
Thanks for bringing this up. I agree with your suggestion of warning if the token is going to be ignored. I've added similar style warnings for other cases (such as `secret-key` and `token`) as well.

Here's the message I've added for this case.
```
warning: {sid} has a token configured in {} that will be ignored because the `cargo:token` credential 
provider is not listed in  `registry.global-credential-providers`
```